modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1160 @@
+        _boundary_edges[_boundary_edges.size() - 1]->id2 = node2;
+      }
+      else
+      {
+        // this is a bad edge; remove it from existing edge list
+        for (unsigned int k = 0; k < _boundary_edges.size(); ++k)
+        {
+          if (_boundary_edges[k]->id1 == node1 && _boundary_edges[k]->id2 == node2)
+          {
+            _boundary_edges.erase(_boundary_edges.begin() + k);
+            break;
+          }
+        }
+      }
+    }
+  }
+}
+
+void
+MeshCut3DUserObject::sortBoundaryNodes()
+// sort boundary nodes to be in the right order along the boundary
+{
+  _boundary.clear();
+
+  for (unsigned int i = 0; i < _boundary_edges.size(); ++i)
+  {
+    unsigned int node1 = _boundary_edges[i]->id1;
+    unsigned int node2 = _boundary_edges[i]->id2;
+
+    mooseAssert(_boundary_map.find(node1) != _boundary_map.end(),
+                "_boundary_map does not have this key")
+        mooseAssert(_boundary_map.find(node2) != _boundary_map.end(),
+                    "_boundary_map does not have this key")
+
+            _boundary_map.find(node1)
+                ->second->neighbor.push_back(node2);
+    _boundary_map.find(node2)->second->neighbor.push_back(node1);
+  }
+
+  std::map<int, std::shared_ptr<CutPoint>>::iterator it = _boundary_map.begin();
+  while (it != _boundary_map.end())
+  {
+    if (it->second->neighbor.size() != 2)
+      mooseError("Boundary points in cutter mesh must have exactly two neighbors");
+    ++it;
+  }
+
+  unsigned int node1 = _boundary_edges[0]->id1;
+  unsigned int node2 = _boundary_edges[0]->id2;
+  _boundary.push_back(node1);
+  _boundary.push_back(node2);
+  for (unsigned int i = 0; i < _boundary_edges.size() - 1; ++i)
+  {
+    mooseAssert(_boundary_map.find(node2) != _boundary_map.end(),
+                "_boundary_map does not have this key")
+
+        unsigned int node3 = _boundary_map.find(node2)->second->neighbor[0];
+    unsigned int node4 = _boundary_map.find(node2)->second->neighbor[1];
+
+    if (node3 == node1)
+    {
+      _boundary.push_back(node4);
+      node1 = node2;
+      node2 = node4;
+    }
+    else if (node4 == node1)
+    {
+      _boundary.push_back(node3);
+      node1 = node2;
+      node2 = node3;
+    }
+    else
+      mooseError("Discontinuity in cutter boundary");
+  }
+}
+
+Real
+MeshCut3DUserObject::findDistance(unsigned int node1, unsigned int node2)
+{
+  Node * n1 = _cut_mesh->node_ptr(node1);
+  mooseAssert(n1 != nullptr, "Node is NULL");
+  Point & p1 = *n1;
+  Node * n2 = _cut_mesh->node_ptr(node2);
+  mooseAssert(n2 != nullptr, "Node is NULL");
+  Point & p2 = *n2;
+  Real distance = (p1 - p2).size();
+  return distance;
+}
+
+void
+MeshCut3DUserObject::refineBoundary()
+// if boundary nodes are too sparse, add nodes in between
+{
+  std::vector<int> new_boundary_order(_boundary.begin(), _boundary.end());
+
+  mooseAssert(_boundary.size() >= 2, "Boundary should have at least two nodes");
+  for (int i = _boundary.size() - 2; i >= 0; --i)
```

Comment:
This will give sign comparison mismatch warnings. It's possible to use unsigned here if you just use "i" as your stopping criteria. Do you really need to loop from the bottom though? What about iterators here?